### PR TITLE
Fix multiple k6 instances can connect to one browser instance and run tests concurrently

### PR DESCRIPTION
**Pull request by inancgumus**
## Problem

When we run multi-VU/instance tests, one of the test runs can close a page of its own. Since each test run attaches to all the browser pages (_this is another area we might want to discuss_), this sometimes creates a race between the test runs. Here is a step by step explanation.

There are two instances and different sessions. However, both instances attach/detach from the same pages since we run in the same browser. `instance2` panics while trying to send CDP messages to a closed web socket connection since it detaches from it (because `instance1` closes it) and tries to attach to it concurrently (race).

| Instance | Notes                                                                                               | Instance Attachments | Browser pages |
|----------|-----------------------------------------------------------------------------------------------------|----------------------|---------------|
| instance1| newPage                                                                                             |                      | page1         |
| instance1| attach to page 1                                                                                    | page1                | page1         |
| instance2| newPage                                                                                             | page2                | page2, page1  |
| instance1| attach to page 2                                                                                    | page2, page1 (!)     | page2, page1  |
| instance2| attach to page 1<br>🚨 ...note that this operation continues concurrently...                        | page2, page1...      | page2, page1  |
| instance1| page1.Close()                                                                                       | page2, page1         | page2         |
| instance1| detach from page 1<br>closes page1 session for instance1. these sessions are not shared with other instances. they are specific to this instance. however, the pages on the browser are shared between instances. | page2                | page2         |
| instance2| detach from page 1<br>closes the page1 (targetID) session for instance2.                            | page2                | page2         |
| instance2| attach to page 1 still continues...<br>🚨panic: session does not exist.<br>There was a panic because instance2.attach to page 1 operation was going on before instance1 closed page1.  | page2, page1...      | page2, page1  |

```mermaid
sequenceDiagram
    participant instance1
    participant instance2
    participant Browser pages
    instance1->>Browser pages: newPage
    instance1->>page1: attach to page 1
    instance2->>Browser pages: newPage
    instance1->>page2: attach to page 2
    instance2->>page1: attach to page 1: did not end yet.
    instance1->>page1: page1.Close()
    instance1->>+page1: detach from page 1
    instance2->>+page1: detach from page 1
    instance2->>+page1: attach to page 1 🚨 panic
```

## Fix

[This check](https://github.com/grafana/xk6-browser/pull/848/commits/7d199c260442e3a6bc04233e4c130d98c863fb9c) allows the test run to continue instead of panicking because of another test run's page detachments. The fix allows running the k6 browser in high concurrency with multiple VUs and instances. Note that it is still *possible to get errors since we need to correctly handle the CDP messages* while sending and receiving them (the order of them). This can cause timeout and other sorts of errors.

## Testing

Here's the test script I used to test this:

```js
import { check } from 'k6';
import { chromium } from 'k6/x/browser';

export default async function () {
  const URL = "ws://127.0.0.1:9222/devtools/browser/035490d7-d3d2-4426-9e38-450adcc7cd74";
  
  console.log('Connecting to browser... VU:', __VU, 'ITER:', __ITER);  
  const browser = chromium.connect(URL);

  console.log('Creating new context... VU:', __VU, 'ITER:', __ITER);
  const context = browser.newContext();

  console.log('Opening new page... VU:', __VU, 'ITER:', __ITER);
  const page = context.newPage();

  console.log('Navigating to website... VU:', __VU, 'ITER:', __ITER);
  await page.goto('https://test.k6.io/', { waitUntil: 'networkidle' });

  console.log('Getting page title... VU:', __VU, 'ITER:', __ITER);
  console.log('page title:'+ page.title(), 'VU:', __VU, 'ITER:', __ITER);
  check(page, {
     'title': p => p.title() == 'Demo website for load testing',
  });
  
  console.log('Closing page... VU:', __VU, 'ITER:', __ITER);
  page.close()
  console.log('Closing context... VU:', __VU, 'ITER:', __ITER);
  context.close()
  console.log('Disconnecting from browser... VU:', __VU, 'ITER:', __ITER);
  browser.close()

  console.log('❌ connected : ', isConnected(browser));
  console.log('Test completed.');
}

function isConnected(browser) {
  return browser.isConnected() ? '✅' : '❌';
}
```

Here's the bash script (`multik6b.sh`) that can run multiple tests (a courtesy of @ankur22 🙇):

```bash
#!/bin/bash

NUM_K6_INSTANCES=$1
TEST_ITERATION=$2
TEST_FILE=$3
LOG_TRACE='info'

if [[ ! -f .last_run ]];
then
    echo "Last run    : never"
else
    echo "Last run    : $(<.last_run)"
fi
echo "Current time: $(date '+%Y-%m-%d %H:%M:%S')"
echo

# # Delete log files from previous runs
rm -f log_*.log

# Check if any files have changed since the last run
if [[ ! -f .last_run || $(fd --type f --changed-within "$(<.last_run)" --exclude 'k6' | wc -l) -gt 0 ]]; then
    echo "✅ Files have changed since the last run, rebuilding k6-browser"
    echo "------------------------------------------------------"
    # Rebuild k6-browser if any files have changed
    xk6 build --with github.com/grafana/xk6-browser=.
fi

date '+%Y-%m-%d %H:%M:%S' > .last_run

echo
echo "------------------------------------------------------"
echo

run_test(){
    index=$1
    if XK6_BROWSER_LOG=$LOG_TRACE ./k6 run -q --vus 1 -i $TEST_ITERATION $TEST_FILE > log_$index.log 2>&1; then
        echo "✅ test run $index succeeded"
    else
        echo "❌ test run $index failed"
    fi
}

i=1
while [[ $i -le $NUM_K6_INSTANCES ]]
do
    run_test $i &
    ((i = i + 1))
done

wait
echo .
echo "All instances exited"
```

Here's an example command for testing:

```console
$ ./multik6b.sh 10 2 script.js
Last run    : 2023-04-05 10:59:40
Current time: 2023-04-05 11:00:02

------------------------------------------------------

✅ test run 5 succeeded
✅ test run 6 succeeded
✅ test run 7 succeeded
✅ test run 2 succeeded
✅ test run 4 succeeded
✅ test run 3 succeeded
✅ test run 1 succeeded
✅ test run 10 succeeded
✅ test run 8 succeeded
✅ test run 9 succeeded
```

The PR also refactors the page attachment logic for better maintenance and readability—also fixes the linter warnings. This helped me find the problem since it made it easier to understand the code. I didn't prefer to put it in another PR but rather as a commit here. I believe "make it better than you found it" is a nice approach for reducing technical debt :)

### CODE REVIEW

File: common/browser.go [hunk 1]
```diff
@@ -255,8 +255,17 @@ func (b *Browser) onAttachedToTarget(ev *target.EventAttachedToTarget) {
 					ev.SessionID, evti.TargetID, b.ctx.Err())
 				return // ignore
 			default:
-				k6ext.Panic(b.ctx, "creating a new background page: %w", err)
 			}
+			// If the session is closed, then ignore the error and return early.
+			// This can happen if the page is closed before the attachedToTarget
+			// event is handled. Another VU or instance can close the page.
```
Comment (ankur22) [inline common/browser.go]: Could you clarify how "Another VU or instance can close the page"? I'm assuming it's only ones that are related to the same `session`?

File: common/browser.go [hunk 2]
```diff
@@ -144,6 +144,18 @@ func (b *Browser) disposeContext(id cdp.BrowserContextID) error {
 	return nil
 }
 
+// getBrowserContextForPageByID returns the BrowserContext for the given page ID.
+// If the browser context is not found, the default BrowserContext is returned.
+func (b *Browser) getBrowserContextForPageByID(id cdp.BrowserContextID) *BrowserContext {
```
Comment (ankur22) [inline common/browser.go:149]: This doesn't seem to be related to a page or a page id. Should it be `getBrowserContextByID`?

File: common/browser.go [hunk 3]
```diff
@@ -208,130 +220,133 @@ func (b *Browser) initEvents() error {
 	return nil
 }
 
+// onAttachedToTarget is called when a new page is attached to the browser.
```
Comment (ankur22) [inline common/browser.go:223]: It's a little difficult to follow what's going on here. Do you think you could revert the refactoring (splitting the large method into smaller ones), so that we can see the actual changes to the `onAttachedToTarget` logic, and then have the refactoring in another commit?

**Review by ankur22 — CHANGES_REQUESTED** [review]
Great bit of debugging to catch this and resolve it! Nice work 👏

I'm wondering if you could help me a bit by splitting the last commit into two, one for the new changes to `onAttachedToTarget` and another for the splitting/refactoring the method into smaller methods?

File: common/browser.go (hunk 1 above)
Comment (inancgumus) [reply common/browser.go]: Does the following clarify your confusion? Let me know if it doesn't help :)

I explain this in the PR description.

```
                                          browser pages
instance1 ---> newPage                    page1
  attaches to: page1
instance2
  attaches to: page1
instance2 ---> newPage                    page1, page2
  attaches to: page2
instance1
  attaches to: page2
instance1 --> page1.Close()               page2
instance1 <-- detachedFromTarget(page1)
  closes page1 session.

--- THIS IS THE RACY PART (happens concurrently) ---
instance2 <-- detachedFromTarget(page1)
  closes page1 session.
instance2 ---> attachToTarget(page1)
  panic: session does not exist.

instance2 panics while trying to send CDP messages to a closed web socket connection since it detaches from it and tries to attach to it concurrently.
```

File: common/browser.go (hunk 2 above)
Comment (inancgumus) [reply common/browser.go:149]: How about `getBrowserContextByIDOrDefault`? The reason is this also returns a default browser context if it cannot find the browser context. I'm open to other naming suggestions.

File: common/browser.go (hunk 3 above)
Comment (inancgumus) [reply common/browser.go:223]: Sorry for the mess. You're right. I'll split the commits and let you know.

File: common/browser.go (hunk 1 above)
Comment (inancgumus) [reply common/browser.go]: I'm also writing a test to validate this behavior.

File: common/browser.go (hunk 2 above)
Comment (inancgumus) [reply common/browser.go]: I've changed it to `getDefaultBrowserContextOrByID`, [here](https://github.com/grafana/xk6-browser/pull/848/commits/1a167e79d711901df566f34332c82ba8c0d50c60). WDYT?

File: common/browser.go [hunk 4]
```diff
@@ -256,32 +256,11 @@ func (b *Browser) onAttachedToTarget(ev *target.EventAttachedToTarget) {
...
-			// This can happen if the page is closed before the attachedToTarget
-			// event is handled.
-			if session.Closed() {
-				b.logger.Debugf("Browser:onAttachedToTarget:return:session.Done",
-					"session closed: sid:%v tid:%v pageType:%s err:%v",
-					ev.SessionID, targetPage.TargetID, targetPage.Type, err)
-				return
-			}
-			k6ext.Panic(b.ctx, "creating a new background page: %w", err)
+			k6ext.Panic(b.ctx, "creating a new %s: %w", targetPage.Type, err)
```
Comment (ankur22) [inline common/browser.go]: Could we return a `FatalError` here instead of panicking?

File: common/browser.go [hunk 5]
```diff
@@ -361,6 +361,44 @@ func (b *Browser) isAttachedPageValid(ev *target.EventAttachedToTarget, browserC
...
+
+	// If we're no longer connected to browser, then ignore WebSocket errors.
+	// This can happen when the browser is closed while the page is being attached.
+	var (
+		isRunning = atomic.LoadInt64(&b.state) == BrowserStateOpen && b.IsConnected() // b.conn.isConnected()
+		wsErr     *websocket.CloseError
+	)
+	if !errors.As(err, &wsErr) && !isRunning {
+		// If we're no longer connected to browser, then ignore WebSocket errors
+		b.logger.Debugf("Browser:onAttachedToTarget:return", "sid:%v tid:%v pageType:%s websocket err:%v",
```
Comment (ankur22) [inline common/browser.go:326]:
```suggestion
		b.logger.Debugf("Browser: isPageAttachmentErrorIgnorable:return", "sid:%v tid:%v pageType:%s websocket err:%v",
```

Same for the others in this method.

File: common/browser.go [hunk 6]
```diff
@@ -297,6 +297,26 @@ func (b *Browser) onAttachedToTarget(ev *target.EventAttachedToTarget) {
 	}
 }
 
+// attachNewPage registers the page as an active page and attaches the sessionID with the targetID.
+func (b *Browser) attachNewPage(p *Page, ev *target.EventAttachedToTarget) {
+	targetPage := ev.TargetInfo
+
+	// Register the page as an active page.
+	b.pagesMu.Lock()
+	b.logger.Debugf("Browser:onAttachedToTarget:addTarget", "sid:%v tid:%v pageType:%s",
```
Comment (ankur22) [inline common/browser.go:277]: Possibly pretty minor in terms of efficiency, but could we log the debug logs outside of the lock/unlock block?

File: common/browser.go (hunk 3 above)
Comment (ankur22) [reply common/browser.go:223]: Nicely done. very easy to follow now, thanks!

**Review by ankur22 — APPROVED** [review]
👏 Thanks for splitting the changes into smaller commits, it was a lot easier to follow. I only have some very minor suggestions.

LGTM 🎉

File: common/browser.go (hunk 4 above)
Comment (inancgumus) [reply common/browser.go]: Happy to help, but currently, making that change will require quite a few structural adjustments since it's called from a goroutine. However, let's discuss how to handle this kind of situation and work on it in another PR. Sound good? :)

File: common/browser.go (hunk 5 above)
Comment (inancgumus) [reply common/browser.go:326]: I missed this, thanks. See: 745cf5499.

File: common/browser.go (hunk 6 above)
Comment (inancgumus) [reply common/browser.go:277]: Makes sense 👍  See: 745cf5499

**Review by ka3de — APPROVED** [review]
Great work @inancgumus ! :tada: 
LGTM.
